### PR TITLE
Multiplex LSP server<->compiler over stdout.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

**Comment by floitsch** [comment]
@kasperl please review the C++ code
@jstroem please review the Go code

File: tools/toitlsp/lsp/compiler/file_server_pipe.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+func (s *PipeFileServer) ConfigLine() string {
+	return "-2"
+}
+
+func (s *PipeFileServer) Protocol() *CompilerFSProtocol {
+	return s.cp
+}
+
+func (s *PipeFileServer) Stop() error {
+	s.l.Lock()
```
Comment (jstroem) [inline tools/toitlsp/lsp/compiler/file_server_pipe.go:61]: do:
```
s.l.Lock()
defer s.l.Unlock()
```

File: tools/toitlsp/lsp/compiler/file_server_pipe.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"io"
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+type PipeFileServer struct {
+	l           sync.Mutex
```
Comment (jstroem) [inline tools/toitlsp/lsp/compiler/file_server_pipe.go:26]: What does this lock cover?, if its `r` and `w` you should use the lock in `Run` and `IsReady` as well?

File: tools/toitlsp/lsp/compiler/multiplex.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	return &pipe{
+		r: r,
+		w: w,
+	}
+}
+
+func (p *pipe) Close() error {
+	err_read := p.r.Close()
+	err_write := p.w.Close()
+	if err_read == nil {
```
Comment (jstroem) [inline tools/toitlsp/lsp/compiler/multiplex.go:119]: I would prefer:
```
if err_read != nil {
  return err_read
}
return err_write
```

**Review by jstroem — APPROVED** [review]

File: tools/toitlsp/lsp/compiler/file_server_pipe.go (hunk 2 above)
Comment (floitsch) [reply tools/toitlsp/lsp/compiler/file_server_pipe.go]: done.

File: tools/toitlsp/lsp/compiler/file_server_pipe.go (hunk 1 above)
Comment (floitsch) [reply tools/toitlsp/lsp/compiler/file_server_pipe.go:64]: done.

File: tools/toitlsp/lsp/compiler/multiplex.go (hunk 3 above)
Comment (floitsch) [reply tools/toitlsp/lsp/compiler/multiplex.go]: done.

File: src/compiler/compiler.cc [hunk 4]
```diff
@@ -331,35 +332,39 @@ void Compiler::language_server(const Compiler::Configuration& compiler_config) {
   Filesystem* fs = null;
   LspFsProtocol* fs_protocol = null;
   LspFsConnection* connection = null;
+  LspWriter* writer;
```
Comment (kasperl) [inline src/compiler/compiler.cc:335]: The initialization code below is complex enough that I'd want to initialize this to null here.

File: src/compiler/lsp/multiplex_stdout.cc [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+void LspWriterMultiplexStdout::printf(const char* format, va_list& arguments) {
+  va_list copy;
+  va_copy(copy, arguments);
+  int32 needed_bytes = static_cast<int32>(vsnprintf(null, 0, format, arguments));
+  checked_fwrite(&needed_bytes, sizeof(needed_bytes));
+  int written_bytes = vprintf(format, copy);
+  if (written_bytes < 0) {
+    perror("multi-printf");
```
Comment (kasperl) [inline src/compiler/lsp/multiplex_stdout.cc:37]: Why multi?

Comment (kasperl) [reply src/compiler/lsp/multiplex_stdout.cc:37]: I typically use something like perror("LspWriterMultiplexStdout/vprintf") to indicate which function failed and where the call was from-ish.

File: src/compiler/lsp/multiplex_stdout.cc [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+void LspWriterMultiplexStdout::write(const uint8* data, int size) {
+  int32 size32 = static_cast<int32>(size);
+  checked_fwrite(&size32, sizeof(size32));
+  checked_fwrite(data, size);
+}
+
+void LspFsConnectionMultiplexStdout::putline(const char* line) {
+  int len = static_cast<int>(strlen(line));
+  int32 size = static_cast<int32>(len) + 1; // +1 for the newline.
+  // Mark as cming from the FS protocol:
```
Comment (kasperl) [inline src/compiler/lsp/multiplex_stdout.cc:54]: cming -> coming

File: src/compiler/lsp/multiplex_stdout.cc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  if (written_char != '\n') FATAL("Couldn't write newline");
+  fflush(stdout);
+}
+
+char* LspFsConnectionMultiplexStdout::getline() {
+  // TODO(florian): we should never need that much.
+  const int MAX_LINE_SIZE = 64 * 1024;
+  char buffer[MAX_LINE_SIZE];
+  // Add a marker to make sure we don't run out of space in the line.
+  buffer[MAX_LINE_SIZE - 1] = 'a';
```
Comment (kasperl) [inline src/compiler/lsp/multiplex_stdout.cc:68]: Maybe pick something less likely to be in the input?

**Review by kasperl — APPROVED** [review]
C++ looks good to me.

File: src/compiler/compiler.cc (hunk 4 above)
Comment (floitsch) [reply src/compiler/compiler.cc]: done.

File: src/compiler/lsp/multiplex_stdout.cc (hunk 5 above)
Comment (floitsch) [reply src/compiler/lsp/multiplex_stdout.cc]: Used `LspWriterMultiplexStdout::printf/vprintf`

File: src/compiler/lsp/multiplex_stdout.cc (hunk 6 above)
Comment (floitsch) [reply src/compiler/lsp/multiplex_stdout.cc]: done.

File: src/compiler/lsp/multiplex_stdout.cc (hunk 7 above)
Comment (floitsch) [reply src/compiler/lsp/multiplex_stdout.cc]: Changed to '@', but anything is ok, as `fgets` is guaranteed to terminate the line with '\0'.
Added comment.

**Pull request merged**